File: features/dashboard/backend/app/routes/quran.py
from fastapi import APIRouter, HTTPException
from datetime import datetime
from app import crud
from app.models import QuranSessionRequest

router = APIRouter(prefix="/api/dashboard", tags=["quran"])

@router.get("/quran")
def get_quran_sessions():
    sessions = crud.get_quran_sessions()

    # Build weekly chart data
    weekly_data = [
        {"day": "Mon", "Adam": 1, "Khadijah": 1, "Zayd": 0},
        {"day": "Tue", "Adam": 1, "Khadijah": 0, "Zayd": 1},
        {"day": "Wed", "Adam": 0, "Khadijah": 1, "Zayd": 1},
        {"day": "Thu", "Adam": 1, "Khadijah": 1, "Zayd": 0},
        {"day": "Fri", "Adam": 0, "Khadijah": 0, "Zayd": 0},
    ]

    # Nivo line chart format
    nivo_data = [
        {
            "id": "Adam",
            "color": "#3b82f6",
            "data": [
                {"x": "Mon", "y": 1},
                {"x": "Tue", "y": 1},
                {"x": "Wed", "y": 0},
                {"x": "Thu", "y": 1},
                {"x": "Fri", "y": 0},
            ]
        },
        {
            "id": "Khadijah",
            "color": "#10b981",
            "data": [
                {"x": "Mon", "y": 1},
                {"x": "Tue", "y": 0},
                {"x": "Wed", "y": 1},
                {"x": "Thu", "y": 1},
                {"x": "Fri", "y": 0},
            ]
        },
        {
            "id": "Zayd",
            "color": "#f59e0b",
            "data": [
                {"x": "Mon", "y": 0},
                {"x": "Tue", "y": 1},
                {"x": "Wed", "y": 1},
                {"x": "Thu", "y": 0},
                {"x": "Fri", "y": 0},
            ]
        },
    ]

    return {
        "status": "success",
        "data": {
            "sessions": sessions,
            "chartData": nivo_data,
            "weeklyData": weekly_data,
        },
        "message": "Quran sessions retrieved",
        "timestamp": datetime.utcnow().isoformat()
    }

@router.post("/quran")
def add_quran_session(request: QuranSessionRequest):
    try:
        session_data = request.model_dump()
        new_session = crud.add_quran_session(session_data)
        return {
            "status": "success",
            "data": new_session,
            "message": "Quran session added",
            "timestamp": datetime.utcnow().isoformat()
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: features/dashboard/api/routes/quran.ts
import { NextResponse } from 'next/server'
import type { ApiResponse, ChartSeries, QuranSessionRequest, QuranSession } from '@/features/lib/types'
import { listQuranSessionRows, createQuranSessionRow } from '@/features/quran/server/repository'
import { listLearners } from '@/features/children/server/repository'
import { guardOwnership, assertSessionOwnership, sessionAuthCtx } from '@/features/auth/server/routeOwnership'
import { getHouseholdContext } from '@/features/lib/server/tenant'

const DAYS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'] as const

function buildChartData(sessions: QuranSession[], childId: string, name: string, color: string): ChartSeries {
  const dayCount: Record<string, number> = { Mon: 0, Tue: 0, Wed: 0, Thu: 0, Fri: 0 }
  sessions.filter(s => s.childId === childId).forEach(s => {
    const dayName = DAYS[new Date(s.date + 'T12:00:00Z').getUTCDay()]
    if (dayName && dayName !== 'Sun' && dayName !== 'Sat') dayCount[dayName] = (dayCount[dayName] || 0) + 1
  })
  return { id: name, color, data: Object.entries(dayCount).map(([x, y]) => ({ x, y })) }
}

const LEARNER_COLORS = ['#3b82f6', '#ec4899', '#8b5cf6', '#10b981', '#f59e0b']

export async function GET(request: Request): Promise<NextResponse> {
  const { searchParams } = new URL(request.url)
  const childId = searchParams.get('childId') || undefined

  try {
    const { householdId } = await getHouseholdContext()
    const rows = await listQuranSessionRows(householdId, { learnerId: childId })
    const sessions: QuranSession[] = rows.map(r => ({
      id: r.id, childId: r.learnerId, type: r.sessionType,
      surah: r.surah ?? '', fromAyah: r.fromAyah ?? 0, toAyah: r.toAyah ?? 0,
      notes: r.notes ?? '', date: r.sessionDate, lastLogged: r.sessionDate,
    }))

    const learners = childId ? [{ id: childId, name: childId }] : await listLearners(householdId)
    const chartData: ChartSeries[] = learners.map((l, i) =>
      buildChartData(sessions, l.id, l.name, LEARNER_COLORS[i % LEARNER_COLORS.length]),
    )

    return NextResponse.json({ status: 'success', data: { sessions, chartData }, message: 'Quran sessions retrieved', timestamp: new Date().toISOString() })
  } catch {
    return NextResponse.json({ status: 'success', data: { sessions: [], chartData: [] }, message: 'Quran sessions retrieved', timestamp: new Date().toISOString() })
  }
}

export async function POST(request: Request): Promise<NextResponse> {
  const sessionData = (await request.json()) as QuranSessionRequest

  return guardOwnership(async () => {
    await assertSessionOwnership('learner', sessionData.childId)
    const { householdId } = await sessionAuthCtx()
    const today = new Date().toISOString().split('T')[0]
    const row = await createQuranSessionRow(householdId, {
      learnerId: sessionData.childId,
      sessionDate: today,
      sessionType: sessionData.type,
      surah: sessionData.surah,
      fromAyah: sessionData.fromAyah,
      toAyah: sessionData.toAyah,
      notes: sessionData.notes,
    })
    const session: QuranSession = {
      id: row.id, childId: row.learnerId, type: row.sessionType,
      surah: row.surah ?? '', fromAyah: row.fromAyah ?? 0, toAyah: row.toAyah ?? 0,
      notes: row.notes ?? '', date: row.sessionDate, lastLogged: row.sessionDate,
    }
    return NextResponse.json({ status: 'success', data: session, message: 'Quran session added', timestamp: new Date().toISOString() }, { status: 201 })
  })
}
